### ISSUE TITLE
whitespace between indices for array in namelist cannot be parsed by nag

### ISSUE THREAD

**Issue by gitporst** [marshallward/f90nml#75]
When writing namelists containing nested structures that need to be indexed, e.g.
```&
idx_nml
   v(1, 1) = 5
   v(2, 1) = 5
/
```

the resulting namelist file cannot be parsed by a program compiled with nag (tested here with version 6.1) , if there is a whitespace between the indices. I was wondering, if that whitespace can be omitted?

**Comment by marshallward** [comment]
Wow, that is shocking.  So NAG Fortran simply cannot read this namelist?

I can add a property to control this.  I guess the only issue is whether to restrict this to index, or to set a general comma spacing rule.

**Comment by gitporst** [comment]
I think everything left of the equal operator should not contain any whitespaces.

**Comment by marshallward** [comment]
I think I will keep the current spacing after the comma as default, if only because I can't find any style guides which suggest otherwise, but I have coded something up which allows the user to remove the spaces, and which seems to work well.  I will try to push this change once I am back at work.

I think the preferred default behaviour is certainly something worth discussing, though it might ultimately come down to personal preference.  However, this is *really* something that NAG ought to fix.

**Comment by ThemosTsikas** [comment]
Hello, 

According to the semi-official (the official specification of Fortran 2008 can be purchased from ISO, https://www.iso.org/store.html using code ISO/IEC 1539:2010) Fortran 2008 document (my emphasis in bold):

[10-007r1.pdf](https://github.com/marshallward/f90nml/files/2022785/10-007r1.pdf)
https://j3-fortran.org/doc/year/10/10-007r1.pdf

> Section 10.11.2
> 
> The characters in one or more namelist records constitute a sequence of name-value subsequences, each of which consists of an object designator followed by an equals and followed by one or more values and value separators. **The equals may optionally be preceded or followed by one or more contiguous blanks.** The end of a record has the same effect as a blank character, unless it is within a character constant. Any sequence of two or more consecutive blanks is treated as a single blank, unless it is within a character constant.
> 
> Each object designator shall begin with a name from the namelist-group-object-list (5.6) and shall follow the syntax of designator (R601). It shall not contain a vector subscript or an image-selector and shall not designate a zero-sized array, a zero-sized array section, or a zero-length character string. Each subscript, stride, and substring range expression shall be an optionally signed integer literal constant with no kind type parameter specified. If a section subscript list appears, the number of section subscripts shall be equal to the rank of the object. If the namelist group object is of derived type, the designator in the input record may be either the name of the variable or the designator of one of its components, indicated by qualifying the variable name with the appropriate component name. Successive qualifications may be applied as appropriate to the shape and type of the variable represented. **Each designator may be preceded and followed by one or more optional blanks but shall not contain embedded blanks.**
> 

So the 
```
idx_nml
   v(1, 1) = 5
   v(2, 1) = 5
/
```
is not a standard Fortran object designator for the purposes of NAMELIST formatting and the NAG Fortran Compiler adheres to the standard by rejecting it.

**Comment by marshallward** [comment]
I thought the designator in this case was each element, e.g. `x(1:5, 1:7:2)` has the designators `1:5` and `1:7:2` which certainly cannot contain spaces.  But are you saying that `x(1:5, 1:7:2)` as a whole is the designator?

**Comment by marshallward** [comment]
BTW very glad to see a NAG rep give feedback here.

**Comment by ThemosTsikas** [comment]
You are welcome, and yes, the object designator is the entire thing that is "followed by an equals".

**Comment by marshallward** [comment]
Yeah, I guess it's quite clear.  No idea how you found this, but thanks very much for explaining this.  My apologies to you and others at NAG, I have changed the default behaviour to not use spaces.

Having said that, as long as those other charlatan compilers support the spaces, I think I have to continue to do the same :).

**Comment by ThemosTsikas** [comment]
I was directed here by a user of our Compiler asking for support. If you are producing Fortran namelist records, it makes sense to keep to the Standard. We CAN be forgiving sometimes, that is, we can accept violations of the Standard but we do try to keep that to the absolute minimum.

**Comment by gitporst** [comment]
Thank you ThemosTsikas for clarifying the issue and thanks marshallward for implementing this. It now works fine for me.

**Comment by marshallward** [comment]
@ThemosTsikas Apologies for dragging you back into this issue.  I have since checked this issue in four other compilers (gfortran, Intel, Cray, PGI) and they all support the whitespace after the comma.

I fully agree with you that NAG is complying with the standard here, and that the other compilers are incorrect on this issue, but I wonder if I could request that NAG support whitespace between indices, either as default or in some controlled way, if only to conform with the other compilers.

If this can be supported, even if only internally, then it might be possible to request that this constraint be removed from the standard at some time in the future.

**Comment by ThemosTsikas** [comment]
Dear Marshall,

Your enhancement request has been received and will be considered.

Best Regards
Themos Tsikas

**Comment by marshallward** [comment]
Thanks very much Themos.

Also quickly wanted to add that Cray Fortran also supports whitespace after the comma.

I may be able to test against Lahey/Fujitsu as well.
